File: braids/sandbox/backend/handlers/sandbox.go
// Package handlers provides HTTP handlers for sandbox code execution
package handlers

import (
	"encoding/json"
	"net/http"
	"regexp"
	"strings"
	"time"
)

// Supported languages
const (
	LangJavaScript = "javascript"
	LangPython     = "python"
	LangGo         = "go"
)

// Error types for categorization
const (
	ErrorTypeSyntax   = "syntax"
	ErrorTypeRuntime  = "runtime"
	ErrorTypeLogic    = "logic"
	ErrorTypeTimeout  = "timeout"
	ErrorTypeEdgeCase = "edge-case"
)

// RunRequest represents a code execution request
type RunRequest struct {
	Code     string `json:"code"`
	Language string `json:"language"`
	Input    string `json:"input,omitempty"`
}

// RunResponse represents code execution result
type RunResponse struct {
	Success     bool   `json:"success"`
	Output      string `json:"output"`
	Error       string `json:"error,omitempty"`
	ErrorType   string `json:"errorType,omitempty"`
	ExecutionMs int64  `json:"executionMs"`
}

// TestCase represents a single test case
type TestCase struct {
	ID       string      `json:"id"`
	Name     string      `json:"name"`
	Input    interface{} `json:"input"`
	Expected interface{} `json:"expected"`
	Hidden   bool        `json:"hidden"`
	Timeout  int         `json:"timeout,omitempty"`
}

// TestRequest represents a test execution request
type TestRequest struct {
	Code       string     `json:"code"`
	Language   string     `json:"language"`
	ExerciseID string     `json:"exerciseId"`
	TestCases  []TestCase `json:"testCases,omitempty"`
}

// TestResult represents a single test result
type TestResult struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Passed   bool   `json:"passed"`
	Expected string `json:"expected,omitempty"`
	Actual   string `json:"actual,omitempty"`
	Message  string `json:"message,omitempty"`
	Hidden   bool   `json:"hidden"`
}

// TestResponse represents test suite results
type TestResponse struct {
	Success     bool         `json:"success"`
	Passed      int          `json:"passed"`
	Failed      int          `json:"failed"`
	Results     []TestResult `json:"results"`
	ExecutionMs int64        `json:"executionMs"`
	ErrorType   string       `json:"errorType,omitempty"`
}

// SubmitRequest represents a solution submission
type SubmitRequest struct {
	Code             string     `json:"code"`
	Language         string     `json:"language"`
	ExerciseID       string     `json:"exerciseId"`
	TestCases        []TestCase `json:"testCases,omitempty"`
	HintsUsed        int        `json:"hintsUsed"`
	TimeSpentSeconds int        `json:"timeSpentSeconds"`
	ExpectedMinutes  int        `json:"expectedMinutes"`
}

// SubmitResponse represents submission result
type SubmitResponse struct {
	Success      bool         `json:"success"`
	Score        int          `json:"score"`
	Passed       bool         `json:"passed"`
	TestResults  []TestResult `json:"testResults"`
	XPEarned     int          `json:"xpEarned"`
	Achievements []string     `json:"achievements,omitempty"`
	Feedback     string       `json:"feedback,omitempty"`
	ErrorType    string       `json:"errorType,omitempty"`
}

// Blocked patterns for security
var blockedPatterns = map[string][]*regexp.Regexp{
	LangJavaScript: {
		regexp.MustCompile(`(?i)\beval\s*\(`),
		regexp.MustCompile(`(?i)\bFunction\s*\(`),
		regexp.MustCompile(`(?i)\brequire\s*\(`),
		regexp.MustCompile(`(?i)\bimport\s+`),
		regexp.MustCompile(`(?i)\bfetch\s*\(`),
		regexp.MustCompile(`(?i)\bXMLHttpRequest`),
		regexp.MustCompile(`(?i)\bprocess\.`),
	},
	LangPython: {
		regexp.MustCompile(`(?i)\bimport\s+os\b`),
		regexp.MustCompile(`(?i)\bimport\s+subprocess\b`),
		regexp.MustCompile(`(?i)\bimport\s+sys\b`),
		regexp.MustCompile(`(?i)\bfrom\s+os\b`),
		regexp.MustCompile(`(?i)\bexec\s*\(`),
		regexp.MustCompile(`(?i)\beval\s*\(`),
		regexp.MustCompile(`(?i)\bopen\s*\(`),
		regexp.MustCompile(`(?i)\b__import__`),
	},
	LangGo: {
		regexp.MustCompile(`(?i)\bimport\s+"os"`),
		regexp.MustCompile(`(?i)\bimport\s+"net`),
		regexp.MustCompile(`(?i)\bimport\s+"os/exec"`),
		regexp.MustCompile(`(?i)\bimport\s+"io/ioutil"`),
	},
}

// SandboxHandler handles sandbox-related requests
type SandboxHandler struct{}

// NewSandboxHandler creates a new sandbox handler
func NewSandboxHandler() *SandboxHandler {
	return &SandboxHandler{}
}

// HandleRun executes code and returns output
func (h *SandboxHandler) HandleRun(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var req RunRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, RunResponse{
			Success:   false,
			Error:     "Invalid request body",
			ErrorType: ErrorTypeSyntax,
		})
		return
	}

	// Validate language
	if !isValidLanguage(req.Language) {
		writeJSON(w, http.StatusBadRequest, RunResponse{
			Success:   false,
			Error:     "Unsupported language",
			ErrorType: ErrorTypeSyntax,
		})
		return
	}

	// Security check
	if err := checkSecurity(req.Code, req.Language); err != "" {
		writeJSON(w, http.StatusOK, RunResponse{
			Success:   false,
			Error:     err,
			ErrorType: ErrorTypeSyntax,
		})
		return
	}

	// Execute code (simulated for now)
	startTime := time.Now()
	result := simulateExecution(req.Code, req.Language, req.Input)
	result.ExecutionMs = time.Since(startTime).Milliseconds()

	writeJSON(w, http.StatusOK, result)
}

// HandleTest runs code against test cases
func (h *SandboxHandler) HandleTest(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var req TestRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, TestResponse{
			Success:   false,
			ErrorType: ErrorTypeSyntax,
		})
		return
	}

	// Validate
	if !isValidLanguage(req.Language) {
		writeJSON(w, http.StatusBadRequest, TestResponse{
			Success:   false,
			ErrorType: ErrorTypeSyntax,
		})
		return
	}

	// Security check
	if err := checkSecurity(req.Code, req.Language); err != "" {
		writeJSON(w, http.StatusOK, TestResponse{
			Success:   false,
			ErrorType: ErrorTypeSyntax,
		})
		return
	}

	// Run tests
	startTime := time.Now()
	results := runTests(req.Code, req.Language, req.TestCases)
	
	passed := 0
	failed := 0
	var errorType string
	
	for _, r := range results {
		if r.Passed {
			passed++
		} else {
			failed++
			if errorType == "" {
				if r.Hidden {
					errorType = ErrorTypeEdgeCase
				} else {
					errorType = ErrorTypeLogic
				}
			}
		}
	}

	writeJSON(w, http.StatusOK, TestResponse{
		Success:     failed == 0,
		Passed:      passed,
		Failed:      failed,
		Results:     results,
		ExecutionMs: time.Since(startTime).Milliseconds(),
		ErrorType:   errorType,
	})
}

// HandleSubmit processes a solution submission
func (h *SandboxHandler) HandleSubmit(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var req SubmitRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, SubmitResponse{
			Success: false,
		})
		return
	}

	// Validate
	if !isValidLanguage(req.Language) {
		writeJSON(w, http.StatusBadRequest, SubmitResponse{
			Success: false,
		})
		return
	}

	// Security check
	if err := checkSecurity(req.Code, req.Language); err != "" {
		writeJSON(w, http.StatusOK, SubmitResponse{
			Success:   false,
			Score:     0,
			Passed:    false,
			ErrorType: ErrorTypeSyntax,
			Feedback:  "Security check failed",
		})
		return
	}

	// Run tests
	results := runTests(req.Code, req.Language, req.TestCases)
	
	passed := 0
	failed := 0
	var errorType string
	
	for _, r := range results {
		if r.Passed {
			passed++
		} else {
			failed++
			if errorType == "" {
				if r.Hidden {
					errorType = ErrorTypeEdgeCase
				} else {
					errorType = ErrorTypeLogic
				}
			}
		}
	}

	// Calculate score
	score := calculateScore(passed, len(req.TestCases), req.HintsUsed, req.TimeSpentSeconds, req.ExpectedMinutes)
	
	// Calculate XP
	xp := calculateXP(score, failed == 0)
	
	// Generate feedback
	feedback := generateFeedback(passed, failed, score)

	writeJSON(w, http.StatusOK, SubmitResponse{
		Success:     true,
		Score:       score,
		Passed:      failed == 0,
		TestResults: results,
		XPEarned:    xp,
		Feedback:    feedback,
		ErrorType:   errorType,
	})
}

// Helper functions

func isValidLanguage(lang string) bool {
	return lang == LangJavaScript || lang == LangPython || lang == LangGo
}

func checkSecurity(code, language string) string {
	patterns, ok := blockedPatterns[language]
	if !ok {
		return ""
	}

	for _, pattern := range patterns {
		if pattern.MatchString(code) {
			return "Security violation: Blocked pattern detected"
		}
	}

	return ""
}

func simulateExecution(code, language, input string) RunResponse {
	// Simulated execution - in production, this would use actual runners
	var output string

	switch language {
	case LangJavaScript:
		// Look for console.log
		re := regexp.MustCompile(`console\.log\s*\(\s*["']([^"']+)["']\s*\)`)
		matches := re.FindAllStringSubmatch(code, -1)
		for _, m := range matches {
			if len(m) > 1 {
				output += m[1] + "\n"
			}
		}
	case LangPython:
		// Look for print
		re := regexp.MustCompile(`print\s*\(\s*["']([^"']+)["']\s*\)`)
		matches := re.FindAllStringSubmatch(code, -1)
		for _, m := range matches {
			if len(m) > 1 {
				output += m[1] + "\n"
			}
		}
	case LangGo:
		// Look for fmt.Println
		re := regexp.MustCompile(`fmt\.Println\s*\(\s*"([^"]+)"\s*\)`)
		matches := re.FindAllStringSubmatch(code, -1)
		for _, m := range matches {
			if len(m) > 1 {
				output += m[1] + "\n"
			}
		}
	}

	if output == "" {
		output = "[Code executed successfully]\n"
	}

	return RunResponse{
		Success: true,
		Output:  strings.TrimSpace(output),
	}
}

func runTests(code, language string, testCases []TestCase) []TestResult {
	results := make([]TestResult, len(testCases))

	for i, tc := range testCases {
		// Simulated test execution
		// In production, this would actually run the code
		
		// Simple heuristic for demo
		hasFunction := strings.Contains(code, "function") || 
		               strings.Contains(code, "def ") || 
		               strings.Contains(code, "func ")
		hasReturn := strings.Contains(code, "return")
		
		passed := hasFunction && hasReturn
		
		results[i] = TestResult{
			ID:       tc.ID,
			Name:     tc.Name,
			Passed:   passed,
			Hidden:   tc.Hidden,
			Expected: formatValue(tc.Expected),
			Actual:   formatValue(tc.Expected), // Simulated
			Message:  "Test executed",
		}
		
		if !passed {
			results[i].Message = "Test failed - check your implementation"
			results[i].Actual = "[simulated output]"
		}
	}

	return results
}

func formatValue(v interface{}) string {
	b, _ := json.Marshal(v)
	return string(b)
}

func calculateScore(passed, total, hints, timeSpent, expectedMinutes int) int {
	if total == 0 {
		return 0
	}

	// Base score from pass rate
	passRate := float64(passed) / float64(total)
	score := passRate * 100

	// Hint penalty
	hintPenalty := float64(hints * 10)
	if hintPenalty > 30 {
		hintPenalty = 30
	}
	score -= hintPenalty

	// Time bonus
	expectedSeconds := expectedMinutes * 60
	if timeSpent < expectedSeconds/2 {
		score += 10
	}

	// Perfect bonus
	if passRate == 1 && hints == 0 {
		score += 5
	}

	if score < 0 {
		score = 0
	}
	if score > 100 {
		score = 100
	}

	return int(score)
}

func calculateXP(score int, passed bool) int {
	xp := 10 // Base XP

	if passed {
		xp += 40
	}

	xp += (score / 10) * 5

	if score == 100 {
		xp += 25
	}

	return xp
}

func generateFeedback(passed, failed, score int) string {
	if failed == 0 {
		if score == 100 {
			return "🎉 Perfect score! Excellent work!"
		}
		return "✅ All tests passed! Great job!"
	}

	if passed == 0 {
		return "Keep trying! Check your logic and try again."
	}

	return "Almost there! Some tests need attention."
}

func writeJSON(w http.ResponseWriter, status int, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

File: braids/sandbox/frontend/stores/sandbox.ts
// ============================================
// Sandbox Store - Code Execution Engine
// ============================================

import { writable, derived, get } from 'svelte/store';

// ============================================
// Types
// ============================================

export type SupportedLanguage = 'javascript' | 'python' | 'go';
export type ErrorType = 'syntax' | 'runtime' | 'logic' | 'timeout' | 'edge-case';

export interface ExecutionResult {
	success: boolean;
	output: string;
	error?: string;
	errorType?: ErrorType;
	executionMs: number;
}

export interface TestResult {
	id: string;
	name: string;
	passed: boolean;
	expected?: string;
	actual?: string;
	message?: string;
	hidden: boolean;
}

export interface TestSuiteResult {
	success: boolean;
	passed: number;
	failed: number;
	results: TestResult[];
	executionMs: number;
	errorType?: ErrorType;
}

export interface SubmissionResult {
	success: boolean;
	score: number;
	passed: boolean;
	testResults: TestResult[];
	xpEarned: number;
	achievements?: string[];
	feedback?: string;
	errorType?: ErrorType;
}

export interface TestCase {
	id: string;
	name: string;
	input: unknown;
	expected: unknown;
	hidden: boolean;
	timeout?: number;
}

// ============================================
// Constants
// ============================================

const EXECUTION_TIMEOUT = 5000; // 5 seconds
const MAX_OUTPUT_LENGTH = 10240; // 10 KB

// Blocked patterns for security
const BLOCKED_PATTERNS: Record<SupportedLanguage, RegExp[]> = {
	javascript: [
		/\beval\s*\(/gi,
		/\bFunction\s*\(/gi,
		/\brequire\s*\(/gi,
		/\bimport\s+/gi,
		/\bfetch\s*\(/gi,
		/\bXMLHttpRequest/gi,
		/\bprocess\./gi,
		/\b__dirname/gi,
		/\b__filename/gi,
	],
	python: [
		/\bimport\s+os\b/gi,
		/\bimport\s+subprocess\b/gi,
		/\bimport\s+sys\b/gi,
		/\bfrom\s+os\b/gi,
		/\bexec\s*\(/gi,
		/\beval\s*\(/gi,
		/\bopen\s*\(/gi,
		/\b__import__/gi,
	],
	go: [
		/\bimport\s+"os"/gi,
		/\bimport\s+"net/gi,
		/\bimport\s+"os\/exec"/gi,
		/\bimport\s+"io\/ioutil"/gi,
	],
};

// ============================================
// State
// ============================================

interface SandboxState {
	isRunning: boolean;
	lastResult: ExecutionResult | null;
	testResults: TestSuiteResult | null;
	submissionResult: SubmissionResult | null;
}

const initialState: SandboxState = {
	isRunning: false,
	lastResult: null,
	testResults: null,
	submissionResult: null,
};

// ============================================
// Store
// ============================================

function createSandboxStore() {
	const { subscribe, set, update } = writable<SandboxState>(initialState);
	
	return {
		subscribe,
		
		/**
		 * Run code and capture output
		 */
		run: async (code: string, language: SupportedLanguage, input?: string): Promise<ExecutionResult> => {
			update(s => ({ ...s, isRunning: true }));
			
			try {
				// Security check
				const securityError = checkSecurity(code, language);
				if (securityError) {
					const result: ExecutionResult = {
						success: false,
						output: '',
						error: securityError,
						errorType: 'syntax',
						executionMs: 0,
					};
					update(s => ({ ...s, isRunning: false, lastResult: result }));
					return result;
				}
				
				// Execute based on language
				let result: ExecutionResult;
				
				if (language === 'javascript') {
					result = await executeJavaScript(code, input);
				} else {
					// For Python/Go, use simulated execution for now
					result = await simulateExecution(code, language, input);
				}
				
				update(s => ({ ...s, isRunning: false, lastResult: result }));
				return result;
				
			} catch (err) {
				const result: ExecutionResult = {
					success: false,
					output: '',
					error: err instanceof Error ? err.message : 'Unknown error',
					errorType: 'runtime',
					executionMs: 0,
				};
				update(s => ({ ...s, isRunning: false, lastResult: result }));
				return result;
			}
		},
		
		/**
		 * Run code against test cases
		 */
		test: async (
			code: string, 
			language: SupportedLanguage, 
			testCases: TestCase[]
		): Promise<TestSuiteResult> => {
			update(s => ({ ...s, isRunning: true }));
			
			const startTime = performance.now();
			const results: TestResult[] = [];
			let passed = 0;
			let failed = 0;
			let errorType: ErrorType | undefined;
			
			// Security check first
			const securityError = checkSecurity(code, language);
			if (securityError) {
				const result: TestSuiteResult = {
					success: false,
					passed: 0,
					failed: testCases.length,
					results: testCases.map(tc => ({
						id: tc.id,
						name: tc.name,
						passed: false,
						message: 'Security check failed',
						hidden: tc.hidden,
					})),
					executionMs: 0,
					errorType: 'syntax',
				};
				update(s => ({ ...s, isRunning: false, testResults: result }));
				return result;
			}
			
			// Run each test case
			for (const testCase of testCases) {
				try {
					const testResult = await runTestCase(code, language, testCase);
					results.push(testResult);
					
					if (testResult.passed) {
						passed++;
					} else {
						failed++;
						// Determine error type from first failure
						if (!errorType) {
							errorType = testCase.hidden ? 'edge-case' : 'logic';
						}
					}
				} catch (err) {
					failed++;
					results.push({
						id: testCase.id,
						name: testCase.name,
						passed: false,
						message: err instanceof Error ? err.message : 'Test error',
						hidden: testCase.hidden,
					});
					if (!errorType) {
						errorType = 'runtime';
					}
				}
			}
			
			const executionMs = Math.round(performance.now() - startTime);
			
			const result: TestSuiteResult = {
				success: failed === 0,
				passed,
				failed,
				results,
				executionMs,
				errorType,
			};
			
			update(s => ({ ...s, isRunning: false, testResults: result }));
			return result;
		},
		
		/**
		 * Submit solution for scoring
		 */
		submit: async (params: {
			code: string;
			language: SupportedLanguage;
			testCases: TestCase[];
			hintsUsed: number;
			timeSpentSeconds: number;
			expectedMinutes: number;
		}): Promise<SubmissionResult> => {
			update(s => ({ ...s, isRunning: true }));
			
			const { code, language, testCases, hintsUsed, timeSpentSeconds, expectedMinutes } = params;
			
			// Run all tests
			const testResult = await get({ subscribe }).testResults 
				|| await runAllTests(code, language, testCases);
			
			// Calculate score
			const score = calculateScore({
				testsPassed: testResult.passed,
				totalTests: testCases.length,
				hintsUsed,
				timeSpentSeconds,
				expectedMinutes,
			});
			
			// Calculate XP
			const xpEarned = calculateXP(score, testResult.passed === testCases.length);
			
			// Generate feedback
			const feedback = generateFeedback(testResult, score);
			
			const result: SubmissionResult = {
				success: true,
				score,
				passed: testResult.failed === 0,
				testResults: testResult.results,
				xpEarned,
				feedback,
				errorType: testResult.errorType,
			};
			
			update(s => ({ ...s, isRunning: false, submissionResult: result }));
			return result;
		},
		
		/**
		 * Clear all results
		 */
		clear: () => {
			set(initialState);
		},
	};
}

// ============================================
// Execution Helpers
// ============================================

function checkSecurity(code: string, language: SupportedLanguage): string | null {
	const patterns = BLOCKED_PATTERNS[language];
	
	for (const pattern of patterns) {
		if (pattern.test(code)) {
			return `Security violation: Blocked pattern detected. This code pattern is not allowed.`;
		}
	}
	
	return null;
}

async function executeJavaScript(code: string, input?: string): Promise<ExecutionResult> {
	return new Promise((resolve) => {
		const startTime = performance.now();
		let output = '';
		let error: string | undefined;
		let errorType: ErrorType | undefined;
		
		// Create a sandboxed function
		try {
			// Capture console.log output
			const logs: string[] = [];
			const mockConsole = {
				log: (...args: unknown[]) => {
					logs.push(args.map(a => 
						typeof a === 'object' ? JSON.stringify(a) : String(a)
					).join(' '));
				},
				error: (...args: unknown[]) => {
					logs.push('[ERROR] ' + args.map(a => String(a)).join(' '));
				},
				warn: (...args: unknown[]) => {
					logs.push('[WARN] ' + args.map(a => String(a)).join(' '));
				},
			};
			
			// Execute with timeout
			const timeoutId = setTimeout(() => {
				error = 'Execution timeout: Code took too long to execute';
				errorType = 'timeout';
				resolve({
					success: false,
					output: logs.join('\n').slice(0, MAX_OUTPUT_LENGTH),
					error,
					errorType,
					executionMs: EXECUTION_TIMEOUT,
				});
			}, EXECUTION_TIMEOUT);
			
			// Create sandboxed execution context
			const sandboxedCode = `
				(function(console) {
					"use strict";
					${code}
				})
			`;
			
			// Execute
			const fn = eval(sandboxedCode);
			fn(mockConsole);
			
			clearTimeout(timeoutId);
			output = logs.join('\n').slice(0, MAX_OUTPUT_LENGTH);
			
			resolve({
				success: true,
				output,
				executionMs: Math.round(performance.now() - startTime),
			});
			
		} catch (err) {
			const executionMs = Math.round(performance.now() - startTime);
			
			if (err instanceof SyntaxError) {
				errorType = 'syntax';
				error = `Syntax Error: ${err.message}`;
			} else if (err instanceof TypeError) {
				errorType = 'runtime';
				error = `Type Error: ${err.message}`;
			} else if (err instanceof ReferenceError) {
				errorType = 'runtime';
				error = `Reference Error: ${err.message}`;
			} else {
				errorType = 'runtime';
				error = err instanceof Error ? err.message : 'Unknown error';
			}
			
			resolve({
				success: false,
				output,
				error,
				errorType,
				executionMs,
			});
		}
	});
}

async function simulateExecution(
	code: string, 
	language: SupportedLanguage, 
	input?: string
): Promise<ExecutionResult> {
	// Simulated execution for Python and Go
	// This analyzes the code and returns realistic-looking output
	
	await new Promise(r => setTimeout(r, 100 + Math.random() * 200)); // Simulate execution time
	
	const startTime = performance.now();
	
	// Look for print/fmt.Println statements
	let output = '';
	
	if (language === 'python') {
		const printMatches = code.matchAll(/print\s*\(\s*(.+?)\s*\)/g);
		for (const match of printMatches) {
			const content = match[1];
			// Simple string extraction
			const stringMatch = content.match(/["'](.+?)["']/);
			if (stringMatch) {
				output += stringMatch[1] + '\n';
			} else if (/^\d+$/.test(content.trim())) {
				output += content.trim() + '\n';
			} else {
				// Variable or expression - just show placeholder
				output += `[Output of ${content.trim()}]\n`;
			}
		}
	} else if (language === 'go') {
		const fmtMatches = code.matchAll(/fmt\.Println\s*\(\s*(.+?)\s*\)/g);
		for (const match of fmtMatches) {
			const content = match[1];
			const stringMatch = content.match(/"(.+?)"/);
			if (stringMatch) {
				output += stringMatch[1] + '\n';
			} else {
				output += `[Output of ${content.trim()}]\n`;
			}
		}
	}
	
	if (!output) {
		output = `[${language} execution simulated - output not captured]\n`;
		output += `Note: Real ${language} execution coming soon!\n`;
	}
	
	return {
		success: true,
		output: output.trim(),
		executionMs: Math.round(performance.now() - startTime),
	};
}

async function runTestCase(
	code: string, 
	language: SupportedLanguage, 
	testCase: TestCase
): Promise<TestResult> {
	// For JavaScript, we can actually run the test
	if (language === 'javascript') {
		return runJavaScriptTest(code, testCase);
	}
	
	// For other languages, simulate test results
	return simulateTestCase(code, language, testCase);
}

async function runJavaScriptTest(code: string, testCase: TestCase): Promise<TestResult> {
	try {
		// Create a test harness
		const testCode = `
			(function() {
				"use strict";
				${code}
				
				// Run the test
				const input = ${JSON.stringify(testCase.input)};
				const expected = ${JSON.stringify(testCase.expected)};
				
				// Try to find a function to call
				const funcMatch = \`${code}\`.match(/function\\s+(\\w+)/);
				if (funcMatch) {
					const funcName = funcMatch[1];
					const result = eval(funcName + '(...(Array.isArray(input) ? input : [input]))');
					return { result, expected };
				}
				
				// If no function, compare console output
				return { result: undefined, expected };
			})()
		`;
		
		const { result, expected } = eval(testCode);
		
		// Compare results
		const passed = JSON.stringify(result) === JSON.stringify(expected);
		
		return {
			id: testCase.id,
			name: testCase.name,
			passed,
			expected: JSON.stringify(expected),
			actual: JSON.stringify(result),
			hidden: testCase.hidden,
			message: passed ? 'Test passed!' : `Expected ${JSON.stringify(expected)}, got ${JSON.stringify(result)}`,
		};
		
	} catch (err) {
		return {
			id: testCase.id,
			name: testCase.name,
			passed: false,
			message: err instanceof Error ? err.message : 'Test execution error',
			hidden: testCase.hidden,
		};
	}
}

async function simulateTestCase(
	code: string, 
	language: SupportedLanguage, 
	testCase: TestCase
): Promise<TestResult> {
	// Simulate test execution with smart pattern matching
	await new Promise(r => setTimeout(r, 50));
	
	// Check for common patterns that would pass/fail
	const hasFunction = /function|def |func /.test(code);
	const hasLoop = /for|while/.test(code);
	const hasReturn = /return/.test(code);
	
	// Simple heuristic: if code has right structure, 70% chance to pass
	const passChance = hasFunction && hasReturn ? 0.8 : hasLoop ? 0.6 : 0.3;
	const passed = Math.random() < passChance;
	
	return {
		id: testCase.id,
		name: testCase.name,
		passed,
		expected: JSON.stringify(testCase.expected),
		actual: passed ? JSON.stringify(testCase.expected) : '[simulated output]',
		hidden: testCase.hidden,
		message: passed 
			? 'Test passed!' 
			: `Simulated: Expected ${JSON.stringify(testCase.expected)}`,
	};
}

async function runAllTests(
	code: string, 
	language: SupportedLanguage, 
	testCases: TestCase[]
): Promise<TestSuiteResult> {
	const results: TestResult[] = [];
	let passed = 0;
	let failed = 0;
	const startTime = performance.now();
	
	for (const tc of testCases) {
		const result = await runTestCase(code, language, tc);
		results.push(result);
		if (result.passed) passed++;
		else failed++;
	}
	
	return {
		success: failed === 0,
		passed,
		failed,
		results,
		executionMs: Math.round(performance.now() - startTime),
	};
}

// ============================================
// Scoring & XP
// ============================================

function calculateScore(params: {
	testsPassed: number;
	totalTests: number;
	hintsUsed: number;
	timeSpentSeconds: number;
	expectedMinutes: number;
}): number {
	const { testsPassed, totalTests, hintsUsed, timeSpentSeconds, expectedMinutes } = params;
	
	if (totalTests === 0) return 0;
	
	// Base score from test pass rate
	const passRate = testsPassed / totalTests;
	let score = passRate * 100;
	
	// Hint penalty: -10% per hint, max -30%
	const hintPenalty = Math.min(hintsUsed * 10, 30);
	score -= hintPenalty;
	
	// Time bonus: +10% if under expected time
	const expectedSeconds = expectedMinutes * 60;
	if (timeSpentSeconds < expectedSeconds * 0.5) {
		score += 10;
	}
	
	// Perfect bonus
	if (passRate === 1 && hintsUsed === 0) {
		score += 5;
	}
	
	return Math.max(0, Math.min(100, Math.round(score)));
}

function calculateXP(score: number, passed: boolean): number {
	// Base XP for attempting
	let xp = 10;
	
	// Bonus for passing
	if (passed) {
		xp += 40;
	}
	
	// Score-based bonus
	xp += Math.floor(score / 10) * 5;
	
	// Perfect score bonus
	if (score === 100) {
		xp += 25;
	}
	
	return xp;
}

function generateFeedback(testResult: TestSuiteResult, score: number): string {
	if (testResult.failed === 0) {
		if (score === 100) {
			return "🎉 Perfect score! Excellent work!";
		}
		return "✅ All tests passed! Great job!";
	}
	
	if (testResult.passed === 0) {
		return "Keep trying! Check your logic and try again.";
	}
	
	const failedTests = testResult.results.filter(r => !r.passed && !r.hidden);
	if (failedTests.length > 0) {
		return `Almost there! ${failedTests.length} test(s) need attention.`;
	}
	
	return "Good progress! Some edge cases need work.";
}

// ============================================
// Export Store
// ============================================

export const sandbox = createSandboxStore();

// ============================================
// Derived Stores
// ============================================

export const isExecuting = derived(sandbox, $s => $s.isRunning);
export const lastOutput = derived(sandbox, $s => $s.lastResult?.output ?? '');
export const lastError = derived(sandbox, $s => $s.lastResult?.error);
export const testsPassed = derived(sandbox, $s => $s.testResults?.passed ?? 0);
export const testsFailed = derived(sandbox, $s => $s.testResults?.failed ?? 0);

